### PR TITLE
http2: compatibility fixes, added docs

### PR DESCRIPTION
**Pull request by mcollina**
I added all the docs (to the best of my knowledge) for the Compatibility API.
I also made `writeHead` behave similarly between HTTP/1 and HTTP/2, and I removed flushHeaders because it was no longer needed after the change.

cc @sebdeckers

### CODE REVIEW

**Comment by sebdeckers** [comment]
@mcollina `res.flushHeaders` is not documented but does exist as an HTTP/1 API. Various tests cover this feature. Wonder if this will break expected behaviour.

https://github.com/nodejs/node/blob/69fdb47e207f9201f36593ba72440e122d193743/lib/_http_outgoing.js#L879

Excellent work on the docs! 👏

**Comment by mcollina** [comment]
@sebdeckers I would prefer not to have it as it is undocumented. If people are using it, we should document it in HTTP/1 as well, and support it. My guess is that it is some leftover from the past

To the best of your knowledge, is it used in Express or similar?

**Comment by mcollina** [comment]
@sebdeckers it's documented for the client request object (https://nodejs.org/api/http.html#http_request_flushheaders), but not for the server response. test-http-flush-response-headers.js covers it for the server response.

@jasnell what do you think?

File: doc/api/http2.md [hunk 1]
```diff
@@ -16,9 +16,10 @@ in order to use the `'http2'` module.
 
 The Core API provides a low-level interface designed specifically around
 support for HTTP/2 protocol features. It is specifically *not* designed for
-compatibility with the existing [HTTP/1][] module API.
+compatibility with the existing [HTTP/1][] module API. However, the [Compatibility API][] is.
```
Comment (jasnell) [inline doc/api/http2.md]: long line?

File: doc/api/http2.md [hunk 2]
```diff
@@ -1171,6 +1174,17 @@ server.on('stream', (stream, headers, flags) => {
...
+#### Event: 'request'
+<!-- YAML
+added: REPLACEME
+-->
+
+* `request` {http2.Http2ServerRequest}
+* `response` {http2.Http2ServerResponse}
+
+Emitted each time there is a request. Note that there may be multiple requests
+per session. See the [Compatibility API](compatiblity-api).
```
Comment (jasnell) [inline doc/api/http2.md]: If we are going to list these here, I think we need to be more explicit about these being part of the compatibility API and calling out the differences between this and the `'stream'` event.

File: doc/api/http2.md [hunk 3]
```diff
@@ -1701,16 +1727,755 @@ req.end('Jane');
...
+The `'request'` event works identically on both [HTTPs](https) and
+HTTP/2.
+
+### Class: http2.Http2ServerRequest
+<!-- YAML
+added: REPLACEME
+-->
+
+A `Http2ServerRequest` object is created by [`http2.Server`][] or
+[`http2.SecureServer`][] and passed as the first argument to the [`'request'`][] event. It may be used to access a request status,
```
Comment (jasnell) [inline doc/api/http2.md]: long line

File: doc/api/http2.md [hunk 4]
```diff
@@ -1701,16 +1727,755 @@ req.end('Jane');
...
+added: REPLACEME
+-->
+
+* `statusCode` {number}
+* `statusMessage` {string}
+* `headers` {Object}
+
+Sends a response header to the request. The status code is a 3-digit HTTP
+status code, like `404`. The last argument, `headers`, are the response headers.
+For compatibility with [HTTP/1](), one can give a human-readable `statusMessage` as the second argument, which will be silenty ignored and emit a warning.
```
Comment (jasnell) [inline doc/api/http2.md]: long line... will stop pointing these out :-)

**Comment by mcollina** [comment]
> One nit... this should land against the initial-pr branch in the nodejs/http2 repo, not my jasnell/node repo. I'll keep this repo in sync for the PR, but let's direct the work over there.

I would need to cherry pick the commits then, the two branches diverged too much.

**Comment by mcollina** [comment]
@sebdeckers I readded `flushHeaders`

**Comment by mcollina** [comment]
@jasnell I've fixed the line length.

File: doc/api/http2.md (hunk 2 above)
Comment (mcollina) [inline doc/api/http2.md]: I think this is great progress, I would prefer landing this, and then we can check where things can be updated. Having some docs for the compat API would help.

**Pull request closed without merging**